### ISSUE TITLE
Define a serialization format, establish socket connections

### ISSUE THREAD

**Issue by amiller** [initc3/HoneyBadgerMPC#59]
We still need to define how parties on different machines actually communicate - right now we've been using an abstract message passing style (placing data as python objects in queues). Now we need to serialize/deserialize them over sockets.

One approach is to use a generic serialization format like Pickle. This is the easiest and has the benefit of being easily extended. Otherwise we can pattern match on session-ID and run a custom decoder for each message type (this is also extensible). Another alternative is Protocol Buffers, which is also extensible but requires a bit more effort (it is also faster and portable across languages).

Using serialization and sockets from honeybadgerbft can be a good starting point. Tom has already built up an implementation in the AVSS implementation https://github.com/tyurek/asyncmix/blob/master/HelperClasses.py
 especially the `Sender`, `Listener` class, which associate an IP/port with each node and serialize/deserialize python objects. The `Config` class uses json but could be adapted to something else. Viff has some serialization but it is a bit more buried/coupled in the overall framework and rely heavily on Twisted. We would need to do something similar using sockets in asyncio.

This can be quite simple. For example, here's code I wrote for honeybadgerBFT that does all the serialization and socket establishing in just ~250 lines of code. 
https://github.com/initc3/HoneyBadgerBFT-Python/blob/demo3/run_fifo.py
The code from Tom could also be ported over fairly directly https://github.com/tyurek/asyncmix/blob/master/TestAvssMultipleProcesses.py
Something like zmq would also be fine.

Solving this problem in some way is the main obstacle for beginning benchmarking of online phase experiments (necessary for #30 among other things)

**Comment by amiller** [comment]
Adding some more thoughts to implementing sockets using TLS for authentication....

Conceptually what we want to do is "mutual authentication". Each of the nodes will have a private key (a `certdata/player-$myid.key)`, and a self-signed certificate `certdata/player-$id.cer`. We all have to agree on this set of certificates, but not otherwise have a CA.
The `generate-certificates.py` module adapted from Viff is now setup to run, and by default puts 4 self-signed certificates in `certdata/`. This script uses pyopenssl, the alternative is just to use openssl from the comment line for this:

      openssl req -x509 -newkey rsa:2048 -keyout ./certdata/player-$myid.key -nodes \
                  -out ./certdata/player-$myid.cert -sha256 -days 1000

The python standard library has support for TLS authentication.
https://docs.python.org/3/library/ssl.html

The basic workflow is you create an `ssl.SSLContext` object, and configure it to load the certificates. This example I believe will set up 
```python
sc = ssl.create_default_context(ssl.Purpose.CLIENT_AUTH, cafile='player-$theirid.cert')
sc.verify_mode(ssl.CERT_REQUIRED)
sc.load_cert_chain('certdata/player-$myid.cert', 'certdata/player-$myid.key')
```

The ssl library then has an integration with python sockets, basically once you have `ssl.SSLContext` object, you can wrap a socket with it:
```
secure_sock =sc.wrap_socket(sock, server_side=True)
```
Here's an example doing mutual authentication: https://stackoverflow.com/questions/44343230/mutual-ssl-authentication-in-simple-echo-client-server-python-sockets-ssl-m?answertab=votes#tab-top

However, this simple interface does not not work correctly with non-blocking sockets

Asyncio has an integration with the `ssl` module built in, but only with its high level streams api. You can specify an `ssl.SSLContext` as a parameter to `asyncio.start_server`.
Here is an example using the SSLContext configured for mutual authentication, using Asyncio ssl  
https://github.com/amiller/HoneyBadgerMPC/blob/ipc-auth/asyncio_ssl_example.py

I don't know if there is a way to use the simple socket wrapping interface in asyncio. I think there isn't. 
I will use a separate comment to discuss authentication using other tools so that we don't have to worry about it in our code.

<s>Here are a couple of examples:
https://pymotw.com/3/asyncio/ssl.html
https://gist.github.com/messa/22398173f039d1230e32#file-asyncio_ssl_example-py</s>

**Comment by amiller** [comment]
The above notes on how to use asyncio and ssl in python are great and totally viable (forcing us to use the stream interface notwithstanding) but....
Here's an even better approach https://github.com/square/ghostunnel
https://github.com/square/ghostunnel#full-tunnel-client-plus-server
This is a dedicated tool for setting up bidirectional TLS-authenticated channels, and forwarding them to different ports. This way we can keep the tls authentication out of our codebase and use the same sockets we do now, at the cost of taking on this dependency.

**Comment by smkuls** [comment]
This has been implemented in #72 and #80
